### ISSUE TITLE
"parser stack overflow" in the SQLite library

### ISSUE THREAD

**Issue by servoz** [populse/populse_db#98]
```
Traceback (most recent call last):
  File "/data/Git_Projects/populse_mia/populse_mia/user_interface/data_browser/data_browser.py", line 747, in search_str
    return_list = [getattr(scan, TAG_FILENAME) for scan in generator]
  File "/data/Git_Projects/populse_mia/populse_mia/user_interface/data_browser/data_browser.py", line 747, in <listcomp>
    return_list = [getattr(scan, TAG_FILENAME) for scan in generator]
  File "/data/Git_Projects/populse_db/python/populse_db/database.py", line 958, in filter_documents
    for doc in self.engine.filter_documents(
  File "/data/Git_Projects/populse_db/python/populse_db/engine/sqlite.py", line 909, in filter_documents
    for doc in self._select_documents(
  File "/data/Git_Projects/populse_db/python/populse_db/engine/sqlite.py", line 716, in _select_documents
    self.cursor.execute(sql, where_data)
sqlite3.OperationalError: parser stack overflow

```

The error message indicates a "parser stack overflow" in the SQLite library, suggesting that the SQL query being executed is too nested or complex to be handled by SQLite.

Looking at the SQL query, it appears to be a SELECT statement with a long list of columns to select and a WHERE clause with multiple conditions, including several LIKE clauses.

Given the complexity of the query and the possibility that it will grow as more conditions are added, we may need to rethink our approach to database querying? There may be another reason I can't see yet (I've just discovered the bug)?

**To reproduce:**

- convert the anat in the Philipps repertory of the [miadatausers](https://gricad-gitlab.univ-grenoble-alpes.fr/condamie/miadatausers) library:
![AnatFromMiadatausers](https://github.com/populse/populse_db/assets/37933899/05838db2-de0a-4cd3-b6b5-bbe4678d1e9e)

- type "a" in the field to filter visible documents:
![filter_orig](https://github.com/populse/populse_db/assets/37933899/e18b0fa3-4791-4610-ae0e-2701d3a57f5e)

-> everything works perfectly!

- click on the "View tags" button and select 30 tags to display:
![30tags](https://github.com/populse/populse_db/assets/37933899/b34a7bd9-aa21-4931-a6ef-433246bf83d3)

- type "a" in the field to filter visible documents:
![tags30](https://github.com/populse/populse_db/assets/37933899/ad554aa9-5177-4e1c-bc37-2f2b4362a52e)

-> everything works perfectly!

- click on the "View tags" button and select at least 35 tags to display:
![tags35](https://github.com/populse/populse_db/assets/37933899/e4fe8bdb-508b-4a6c-ad0f-5c07f3172a29)

- type "a" in the field to filter visible documents:
- Mia crash:
![filterCrash](https://github.com/populse/populse_db/assets/37933899/2a7881c8-edda-42a9-93da-cd69f323ae6b)

The crash comes from the:
```
File "/data/Git_Projects/populse_db/python/populse_db/engine/sqlite.py", line 716, in _select_documents
    self.cursor.execute(sql, where_data)
```
Where:
- **self.cursor** is a sqlite3.Cursor object
- **sql** is  "`SELECT` [!File!Name],[!File!Name],[!Checksum],[!Type],[!Exp !Type],[!History],[!Full history],[!Creation!Date],[!Protocol!Name],[!Direction!Diffusion],[3rd row affine transform],[!Magic filetype string],[!Intent parameters],[!Image!Type],[!Datatype],[!Json_!Version],[!Image!Orientation!Patient],[!Label!Type],[!Intent name],[!Size of header],[!Auxiliary file],[!Scan!Mode],[!Data!Anonymized],[!Grid spacings (!X,!Y,!Z,!T,...)],[!Affine transform code],[!Patient!Name],[!F!O!V],[!Acquisition!Time],[!Study!Name],[!T offset],[!Intent code],[!Image!In!Acquisition],[!Slice!Thickness!In!Acquisition],[!Qform transform code],[!Description],[!Echo!Time],[!Slice duration],[!Endianness],[!Cal. (display) max/min],[!Flip!Angle],[!Slice!Orientation],[!File offset to data blob],[!Start/end slice],[!Acquisition!Date],[!Quaternion b,c,d params],[!Repetition!Time],[!Sequence!Name],[!Dataset dimensions (!Count, !X,!Y,!Z,!T...)],[!Extension byte 1],[!Q factor],[!T units],[2nd row affine transform],[!Image!Position!Patient],[!Scaling slope and intercept],[!X!Y!Z  units],[1st row affine transform],[!Slice timing code],[!Slice!Gap],[!M!R!I slice ordering (freq, phase, slice index)],[!Quaternion x,y,z shifts],[!Slice!Thickness],[!Bits per voxel] `FROM `[current] `WHERE` ( ( [!File!Name] like '%a%' ) or ( ( [!Type] like '%a%' ) or ( ( [!Exp !Type] like '%a%' ) or ( ( [!Creation!Date] like '%a%' ) or ( ( [!Protocol!Name] like '%a%' ) or ( ( [!Direction!Diffusion] like '%a%' ) or ( ( [3rd row affine transform] like '%a%' ) or ( ( [!Magic filetype string] like '%a%' ) or ( ( [!Intent parameters] like '%a%' ) or ( ( [!Image!Type] like '%a%' ) or ( ( [!Datatype] like '%a%' ) or ( ( [!Json_!Version] like '%a%' ) or ( ( [!Image!Orientation!Patient] like '%a%' ) or ( ( [!Label!Type] like '%a%' ) or ( ( [!Intent name] like '%a%' ) or ( ( [!Size of header] like '%a%' ) or ( ( [!Auxiliary file] like '%a%' ) or ( ( [!Scan!Mode] like '%a%' ) or ( ( [!Data!Anonymized] like '%a%' ) or ( ( [!Grid spacings (!X,!Y,!Z,!T,...)] like '%a%' ) or ( ( [!Affine transform code] like '%a%' ) or ( ( [!Patient!Name] like '%a%' ) or ( ( [!F!O!V] like '%a%' ) or ( ( [!Acquisition!Time] like '%a%' ) or ( ( [!Intent code] like '%a%' ) or ( ( [!Image!In!Acquisition] like '%a%' ) or ( ( [!Slice!Thickness!In!Acquisition] like '%a%' ) or ( ( [!Qform transform code] like '%a%' ) or ( ( [!Description] like '%a%' ) or ( ( [!Echo!Time] like '%a%' ) or ( ( [!Slice duration] like '%a%' ) or ( ( [!Endianness] like '%a%' ) or ( ( [!Cal. (display) max/min] like '%a%' ) or ( ( [!Flip!Angle] like '%a%' ) or ( ( [!Slice!Orientation] like '%a%' ) or ( ( [!File offset to data blob] like '%a%' ) or ( ( [!Acquisition!Date] like '%a%' ) or ( ( [!Quaternion b,c,d params] like '%a%' ) or ( ( [!Repetition!Time] like '%a%' ) or ( ( [!Sequence!Name] like '%a%' ) or ( ( [!Dataset dimensions (!Count, !X,!Y,!Z,!T)] like '%a%' ) or ( ( [!Extension byte 1] like '%a%' ) or ( ( [!Q factor] like '%a%' ) or ( ( [2nd row affine transform] like '%a%' ) or ( ( [!Image!Position!Patient] like '%a%' ) or ( ( [!Scaling slope and intercept] like '%a%' ) or ( ( [1st row affine transform] like '%a%' ) or ( ( [!Slice timing code] like '%a%' ) or ( ( [!Slice!Gap] like '%a%' ) or ( ( [!M!R!I slice ordering (freq, phase, slice index)] like '%a%' ) or ( ( [!Quaternion x,y,z shifts] like '%a%' ) or ( ( [!Slice!Thickness] like '%a%' ) or ( [!Bits per voxel] like '%a%' ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) ) and ( [!File!Name] IN ('data/raw_data/alej170316-IRMFonct_+perfusion-2016-03-17083444-01-T13DSENSE-T1TFE-000425_000.nii') )"
- and **where_data** is []

**Comment by sapetnioc** [comment]
Thank you. But I am looking for the query string that must have been built by populse-mia before its parsing. Just to be sure that it is the parsing that adds the parenthesis.

**Comment by servoz** [comment]
oh I see !

in:
[generator = self.project.session.filter_documents(
                COLLECTION_CURRENT, filter
            )](https://github.com/populse/populse_mia/blob/91de03ced21e214c1a20056760f89d51cd1442ab/populse_mia/user_interface/data_browser/data_browser.py#L742)

**filter** is:
'(({FileName} LIKE "%a%") OR ({Type} LIKE "%a%") OR ({Exp Type} LIKE "%a%") OR ({CreationDate} LIKE "%a%") OR ({ProtocolName} LIKE "%a%") OR ({DirectionDiffusion} LIKE "%a%") OR ({3rd row affine transform} LIKE "%a%") OR ({Magic filetype string} LIKE "%a%") OR ({Intent parameters} LIKE "%a%") OR ({ImageType} LIKE "%a%") OR ({Datatype} LIKE "%a%") OR ({Json_Version} LIKE "%a%") OR ({ImageOrientationPatient} LIKE "%a%") OR ({LabelType} LIKE "%a%") OR ({Intent name} LIKE "%a%") OR ({Size of header} LIKE "%a%") OR ({Auxiliary file} LIKE "%a%") OR ({ScanMode} LIKE "%a%") OR ({DataAnonymized} LIKE "%a%") OR ({Grid spacings (X,Y,Z,T,...)} LIKE "%a%") OR ({Affine transform code} LIKE "%a%") OR ({PatientName} LIKE "%a%") OR ({FOV} LIKE "%a%") OR ({AcquisitionTime} LIKE "%a%") OR ({Intent code} LIKE "%a%") OR ({ImageInAcquisition} LIKE "%a%") OR ({Qform transform code} LIKE "%a%") OR ({Description} LIKE "%a%") OR ({EchoTime} LIKE "%a%") OR ({Slice duration} LIKE "%a%") OR ({Endianness} LIKE "%a%") OR ({Cal. (display) max/min} LIKE "%a%") OR ({FlipAngle} LIKE "%a%") OR ({SliceOrientation} LIKE "%a%") OR ({File offset to data blob} LIKE "%a%") OR ({AcquisitionDate} LIKE "%a%") OR ({Quaternion b,c,d params} LIKE "%a%") OR ({RepetitionTime} LIKE "%a%") OR ({SequenceName} LIKE "%a%") OR ({Dataset dimensions (Count, X,Y,Z,T...)} LIKE "%a%") OR ({Extension byte 1} LIKE "%a%") OR ({Q factor} LIKE "%a%") OR ({2nd row affine transform} LIKE "%a%") OR ({ImagePositionPatient} LIKE "%a%") OR ({Scaling slope and intercept} LIKE "%a%") OR ({1st row affine transform} LIKE "%a%") OR ({Slice timing code} LIKE "%a%") OR ({SliceGap} LIKE "%a%") OR ({MRI slice ordering (freq, phase, slice index)} LIKE "%a%") OR ({Quaternion x,y,z shifts} LIKE "%a%") OR ({Bits per voxel} LIKE "%a%")) AND ({FileName} IN ["data/raw_data/alej170316_test24042018-IRMFonct_+perfusion-2016-03-17083444-02-T13DSENSE-T1TFE-000425_000.nii"])'

**Comment by servoz** [comment]
so ?

**Comment by sapetnioc** [comment]
Clearly populse_db is adding a level of perenthesis for each use of `OR`operator. This is useless here and may be the source of the SQL error. I will have to see how to fix that.

Nevertheless, such a request can only work because SQLite is very permissive with data types. Using `LIKE` on fields that are not string is likely to raise an error on many database. I would recommend to build a shorter query by filtering fields according to their type.

**Comment by servoz** [comment]
Thank you very much!

I'd be grateful if you could take a look at the populse_db side.
I'm going to look at the populse_mia side, in order to build a shorter query by filtering fields by type.

There are several ways of filtering documents in the populse_mia Data Browser. The query here corresponds to the default filtering, i.e. it involves entering a string in the search field in order to display all documents containing this string in the metadata (including the document - Filename if I remember correctly -). That's why the query is quite long.

Anyway, I'll have a look at how to do what you recommend. Do you have any advice for me?

**Comment by sapetnioc** [comment]
It will not be simple to fix it I guess because just removing parenthesis break the tests. I will have to dive into the semantic of the filters and how to handle the priority of operators.

For this default request, at least fields that are not of type string or that contain strings (numbers, dates, etc.) can be filtered out. But things may be a bit complex for list and dictionaries. I believe that we should add a specific method to populse_db such as `full_text_search_query(string_to_search: str, include_fields : list[str] | None, ignore_fields : list[str] | None)` that would build the appropriate query taking into account filed types and database backend.

**Comment by servoz** [comment]
Currently the filter is done in populse_mia.
I'm not against it being done directly in populse_db.

First of all I'm going to look at and understand how to do what you recommend (I'm not a database expert so this will allow me to learn a bit more about the subject before I can do something) in the [populse_mia rapid_search ](https://github.com/populse/populse_mia/blob/3e142a1568a59d5963c67c23aaeed56b29101c81/populse_mia/user_interface/data_browser/rapid_search.py#L93)module.

Then we can always port it to populse_db. I might need some advice ... to be continued ...

As I'm celebrating work tomorrow ... by not working tomorrow, I'll start again on this ticket on Thursday!

**Comment by servoz** [comment]
In fact, the query is constructed in populse_mia in a very simple and basic way. Basically, we take all the "visible" tags in the database and add to a query string for each tag: `tag LIKE "string to search"` ; then we add to the query string: ` AND FileName IN [documents in database]` :

        query = "("
        or_to_write = False

        for tag in tags :
            if tag != TAG_BRICKS :
                if or_to_write :
                    query += " OR "

                query += "({" + tag + '} LIKE "%" + search + '%")

                or_to_write = True

        query += (
            ") AND ({"
            + TAG_FILENAME
            + "} IN "
            + str(scans).replace("'", '"')
            + ")"
        )
I believe that SQlite only supports a very limited number of types (INTEGER, TEXT, REAL, NUMERIC, and NULL), and in SQLite, it's not mandatory to type queries, as SQLite is a database management system without static typing.

However, as we need structured data, we need to use serialization methods to store structured data in SQLite columns. If memory serves (but I'll have to re-read the code a bit to remember the details), we use the `ast.literal_eval()` function in populse_mia to deserialize a string, returned by the database, that contains a valid Python data structure. I'm not sure if we use json, which can be a good method for serializing and deserializing.

In short, when you talk about filtering by type, you're not talking about the actual query to SQlite. Could you tell us a little more about this? It seems to me that if we want to filter with 1 for example, we have to search for "1" in all the fields, don't we? for example "1" or "[1]" or "{'1':'1'}" ?

**Comment by sapetnioc** [comment]
In populse_db, all field of a collection has a type. Each `tag` in the code above corresponds to a field. If you can have access to a database session and collection_name in the `for` loop, it will be easy to add a type filtering test such as:
```
for tag in tags :
    if tag != TAG_BRICKS and session.get_field(collection_name, tag).field_type in ("string", "json", "list_string", "list_json"):
        ...
```

If the name of the type in the query gives a good guess of the field types, I many fields will be removed by this filter.

I am not completely sure of the syntax because I haven't used this v2 API for years but it gives the idea.

**Comment by servoz** [comment]
Okay, I see.
I'll modify it accordingly.

**Comment by sapetnioc** [comment]
Regarding your last question, I would keep list and dict field types in the query since they are implemented as TEXT in sqlite and, by luck, it will work to search a string in it. But, more generally, we can wonder what is the meaning of a query such as  `"text" LIKE {...}` or `"text" LIKE [...]`. In your case, the meaning is probably "`{...}` contains a string `s` where `"text" LIKE s` is true". But the implementation will not work in all cases. For instance, `"{" LIKE {...}` will always be true whatever is the content of the dictionary. The generated SQL query should depends on the field type and use JSON structure for list and dict.

**Comment by sapetnioc** [comment]
I just checked and, surprisingly, sqlite3 supports the query `SELECT 123 LIKE '%12%'` whereas Postgresql complains:
```
postgres=# select 123 LIKE '%12%';
ERROR:  operator does not exist: integer ~~ unknown
LINE 1: select 123 LIKE '%12%';
                   ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```
Therefore, if you are using the default filter in mia-processes to find numbers, this would work today but would not be possible with type filtering. However, if ever another database is used it would raise an error.

**Comment by servoz** [comment]
I've seen that you've tried to work on deleting parentheses in a branch coming from  3.0. Do you plan to do the same in master? I'm very interested to know if this fixes the original issue!

**Comment by servoz** [comment]
In fact, while working on this issue, I realised that searching for characters in tags (what you might call field filtering...) wasn't really giving the results I thought it would... 

The codes were written several years ago, so I think it would be useful to refresh them.

I'm going to open a ticket on populse_mia side ...

**Comment by servoz** [comment]
I would like to go back over the details of this ticket.
Another ticket has also been opened on Mia's side. I had completely forgotten about that!
I will take care of it as soon as possible.
